File: Data Structures & Algorithms/anagram-groups/submission-7.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # anagram (string) contains the exact same chars
        # group all anagram

        # Solution:
        #   init hashmap
        #   'key' to find anagram group
        #   'key' write a function to hash given string

        def hash_anagram(s) -> str:
            # count fre of char
            # encode them as a string, return

            fres = [0] * 26
            for c in s:
                fres[ord(c) - ord('a')] += 1
            
            res = ''
            for f in fres:
                res += str(f) + ','
            return res
        
        map_anagram = defaultdict(list)
        for s in strs:
            encode = hash_anagram(s)
            print(encode)
            map_anagram[encode].append(s)
        
        res = []
        for key, value in map_anagram.items():
            res.append(value)

        return res

File: Data Structures & Algorithms/longest-palindromic-substring/submission-7.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();

        vector<vector<int>> dp(n, vector<int>(n, 0));

        int res = 0;
        string res_str = "";
        for (int i = n-1; i >= 0; i--) {
            for (int j = i; j < n; j++) {
                dp[i][j] = 1 ? (s[i] == s[j] && (j-i+1 <= 3 || dp[i+1][j-1] == 1)) : 0;
                if (dp[i][j] == 1 && j-i+1 > res) {
                    res = j-i+1;
                    res_str = s.substr(i, j-i+1);
                }
            }
        }

        return res_str;
    }
};
